### PR TITLE
Generic Array types

### PR DESCRIPTION
**Pull request by danpovey**
Guys, this is important.
I want to template-ize the code and make it more general.
This is a draft of how it would look at the C++ level.  Please read ASAP.

### CODE REVIEW

File: k2/csrc/array.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+namespace k2 {
+
+
+/*
+   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and otherwise
+   just T* (which presumably be faster).
+*/
+template <typename I, typename T>
+class StridedPtr {
```
Comment (csukuangfj) [inline k2/csrc/array.h:23]: `struct` or add `public:` inside the class.

File: k2/csrc/array.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  T *data;
+  I stride;
```
Comment (csukuangfj) [inline k2/csrc/array.h:25]: Does the `stride` designate the number of bytes or number of elements?

Comment (danpovey) [reply k2/csrc/array.h:25]: I was thinking number of elements.  We could change it later, of course, if bytes was more convenient.

**Comment by danpovey** [comment]
I pushed a file array.py which shows in broad outline how it would look from Python.

File: k2/csrc/array.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+               // supporting operator [].  data[indexes[0]] through
+               // data[indexes[size] - 1] must be accessible through this
+               // object.
+};
+
+
+template <typename I, typename Ptr>
+struct Array3 {
+  using IndexT = I;
+  using PtrT = Ptr;
```
Comment (csukuangfj) [inline k2/csrc/array.h:55]: duplicate with line 59-60

File: k2/csrc/array.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                           // means the array must be of at least size+1.  We
+                           // require that indexes[i] <= indexes[i+1], but it is
+                           // not required that indexes[0] == 0, it may be
+                           // greater than 0.
+
+  const IndexT *indexes2;  // indexes2[indexes1[0]]
+                           // .. indexes2[indexes1[size]-1] should be defined.
+
+  Ptr data;    // `data` might be an actual pointer, or might be some object
+               // supporting operator [].  data[indexes[0]] through
```
Comment (csukuangfj) [inline k2/csrc/array.h:73]: The comment is not correct.

It should be 

data[ indexes2[ indexes1[0] ] ]
throught
data[ indexes2[ indexes1[size] - 1 ] ]
must be accessible.

File: k2/csrc/array.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+               // object.
+};
+
+
+template <typename I, typename Ptr>
+struct Array3 {
+  using IndexT = I;
+  using PtrT = Ptr;
+
+  // Irregular three dimensional array of something, like vector<vector<X> >
```
Comment (qindazhu) [inline k2/csrc/array.h:57]: vector<vector<vector<X>>> or vector<Array2>?

Comment (qindazhu) [reply k2/csrc/array.h:57]: vector<Array2>

File: k2/csrc/array.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+};
+
+
+// Note: we can create Array4 later if we need it.
+
+
+// we'd put the following in fsa.h
+using Cfsa = Array2<int32_t, Arc>;
```
Comment (qindazhu) [inline k2/csrc/array.h:89]: so will we remove the definition of `Cfsa` and `CfsaVec` in https://github.com/danpovey/k2/blob/d4b99433818021804c43594d1479a574f88bc076/k2/csrc/fsa.h#L140
 as they have the same name.

**Comment by danpovey** [comment]
Guys, please have a look at the latest commit here.
This a draft for how the interfaces of FSA operations would look; they would be wrapped in a class.
This way of doing things allows for easier interaction with multithreaded code and even higher-level arrays; and would, later on, hopefully make it easier to do these kinds of algorithms on a GPU.

**Comment by qindazhu** [comment]
Close this as we have merged it to master by PR https://github.com/danpovey/k2/pull/53

**Pull request closed without merging**